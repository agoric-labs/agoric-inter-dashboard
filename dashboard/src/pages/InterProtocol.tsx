import { useCubeQuery } from '@cubejs-client/react';
import { PieChart, Pie, Cell, Tooltip } from 'recharts';
import { Skeleton } from '@/components/ui/skeleton';
import { SectionHeader } from '@/components/SectionHeader';
import { ValueCard } from '@/components/ValueCard';
import { ValueCardGrid } from '@/components/ValueCardGrid';
import { PageHeader } from '@/components/PageHeader';
import { PageContent } from '@/components/PageContent';
import { colors } from '@/components/palette';
import { formatPercent, roundPrice, formatPrice, formatIST, extractFirstFloat } from '@/utils';
import { ErrorAlert } from '@/components/ErrorAlert';

const firstCards = [
  'IST in Circulation',
  'Total Mint Limit',
  'Total Mint Limit Utilized',
  'Total Interchain IST',
  '% of Interchain IST',
  'Smart Wallets Provisioned',
];

const RADIAN = Math.PI / 180;

const renderCustomizedLabel = ({ cx, cy, midAngle, innerRadius, outerRadius, percent, payload }: any) => {
  const radius = innerRadius + (outerRadius - innerRadius) * 0.4;
  const x = cx + radius * Math.cos(-midAngle * RADIAN);
  const y = cy + radius * Math.sin(-midAngle * RADIAN) - 5;

  return (
    <>
      <text x={x} y={y - 8} fill="white" textAnchor={x > cx ? 'start' : 'end'} dominantBaseline="central">
        {payload.name}
      </text>
      <text x={x} y={y + 8} fill="white" textAnchor={x > cx ? 'start' : 'end'} dominantBaseline="central">
        {`${(percent * 100).toFixed(0)}%`}
      </text>
    </>
  );
};

export function InterProtocol() {
  const wcRes = useCubeQuery({
    measures: ['wallets.address_count'],
  });

  const ibcRes = useCubeQuery({
    measures: ['balances.amount_sum'],
    timeDimensions: [
      {
        dimension: 'balances.day',
        granularity: 'day',
        dateRange: 'from 1 days ago to now',
      },
    ],
    order: { 'balances.day': 'desc' },
    segments: ['balances.interchain_ist'],
  });

  const vgRes = useCubeQuery({
    measures: ['vault_factory_governance.debt_limit_sum'],
    timeDimensions: [
      {
        dimension: 'vault_factory_governance.day',
        granularity: 'day',
        dateRange: 'from 1 days ago to now',
      },
    ],
    order: [['vault_factory_governance.day', 'desc']],
  });

  const vmRes = useCubeQuery({
    measures: ['vault_factory_metrics.total_debt_sum', 'vault_factory_metrics.total_collateral_usd_sum'],
    timeDimensions: [
      {
        dimension: 'vault_factory_metrics.day',
        granularity: 'day',
        dateRange: 'from 1 days ago to now',
      },
    ],
    order: [['vault_factory_metrics.day', 'desc']],
  });

  const psmGovRes = useCubeQuery({
    measures: ['psm_governance.mint_limit_sum'],
    timeDimensions: [
      {
        dimension: 'psm_governance.day',
        granularity: 'day',
        dateRange: 'from 1 days ago to now',
      },
    ],
    order: [['psm_governance.day', 'desc']],
  });

  const psmRes = useCubeQuery({
    measures: ['psm_stats.minted_pool_balance_sum'],
    timeDimensions: [
      {
        dimension: 'psm_stats.day',
        granularity: 'day',
        dateRange: 'from 1 days ago to now',
      },
    ],
    order: [['psm_stats.day', 'desc']],
  });

  const reserveRes = useCubeQuery({
    measures: ['reserve.shortfall_balance_avg'],
    timeDimensions: [{ dimension: 'reserve.day', granularity: 'day' }],
    order: {
      'reserve.day': 'desc',
    },
    limit: 1,
  });

  const reserveAllocRes = useCubeQuery({
    measures: ['reserve_allocations.amount_usd_sum'],
    timeDimensions: [{ dimension: 'reserve_allocations.day', granularity: 'day', dateRange: 'from 1 days ago to now' }],
    order: {
      'reserve_allocations.day': 'desc',
    },
  });

  if (reserveAllocRes.error) {
    return <ErrorAlert value={reserveAllocRes.error} />;
  }

  if (wcRes.error) {
    return <ErrorAlert value={wcRes.error} />;
  }

  if (ibcRes.error) {
    return <ErrorAlert value={ibcRes.error} />;
  }

  if (vmRes.error) {
    return <ErrorAlert value={vmRes.error} />;
  }

  if (vgRes.error) {
    return <ErrorAlert value={vgRes.error} />;
  }

  if (psmGovRes.error) {
    return <ErrorAlert value={psmGovRes.error} />;
  }

  if (psmRes.error) {
    return <ErrorAlert value={psmRes.error} />;
  }

  if (reserveRes.error) {
    return <ErrorAlert value={reserveRes.error} />;
  }

  if (
    wcRes.isLoading ||
    !wcRes.resultSet ||
    ibcRes.isLoading ||
    !ibcRes.resultSet ||
    vmRes.isLoading ||
    !vmRes.resultSet ||
    vgRes.isLoading ||
    !vgRes.resultSet ||
    psmRes.isLoading ||
    !psmRes.resultSet ||
    reserveRes.isLoading ||
    !reserveRes.resultSet ||
    reserveAllocRes.isLoading ||
    !reserveAllocRes.resultSet ||
    psmGovRes.isLoading ||
    !psmGovRes.resultSet
  ) {
    return (
      <>
        <PageHeader title="Summary" />
        <PageContent>
          <ValueCardGrid>
            {firstCards.map(title => (
              <ValueCard title={title} key={title} value={<Skeleton className="w-[100px] h-[32px] rounded-full" />} />
            ))}
          </ValueCardGrid>
          <SectionHeader>Balances</SectionHeader>
          <Skeleton className="w-full h-[20px] rounded-full mb-2" />
          <Skeleton className="w-full h-[20px] rounded-full mb-2" />
          <Skeleton className="w-full h-[20px] rounded-full mb-2" />
          <Skeleton className="w-full h-[20px] rounded-full mb-2" />
        </PageContent>
      </>
    );
  }

  // top cards
  const ibcBalance = extractFirstFloat(ibcRes, 'balances.amount_sum');
  const walletCount = extractFirstFloat(wcRes, 'wallets.address_count');

  const psmMinted = extractFirstFloat(psmRes, 'psm_stats.minted_pool_balance_sum');
  const vaultMinted = extractFirstFloat(vmRes, 'vault_factory_metrics.total_debt_sum');
  const totalMinted = psmMinted + vaultMinted;

  const vaultMintLimit = extractFirstFloat(vgRes, 'vault_factory_governance.debt_limit_sum');
  const psmMintLimit = extractFirstFloat(psmGovRes, 'psm_governance.mint_limit_sum');
  const totalMintLimit = vaultMintLimit + psmMintLimit;

  // bottom cards
  const totalReserve = extractFirstFloat(reserveAllocRes, 'reserve_allocations.amount_usd_sum');
  const reserveShortfall = extractFirstFloat(reserveRes, 'reserve.shortfall_balance_avg');
  const totalLockedCollateral = extractFirstFloat(vmRes, 'vault_factory_metrics.total_collateral_usd_sum');

  return (
    <>
      <PageHeader title="Summary" />
      <PageContent>
        <ValueCardGrid>
          <ValueCard title="IST in Circulation" value={formatIST(totalMinted)} />
          <ValueCard title="Total Mint Limit" value={formatIST(vaultMintLimit + psmMintLimit)} />
          <ValueCard title="Total Mint Limit Utilized" value={formatPercent(totalMinted / totalMintLimit)} />
          <ValueCard title="Total Interchain IST" value={formatIST(ibcBalance)} />
          <ValueCard title="% of Interchain IST" value={formatPercent(ibcBalance / totalMinted)} />
          <ValueCard title="Smart Wallets Provisioned" value={walletCount} />
        </ValueCardGrid>

        <SectionHeader>Balances</SectionHeader>
        <div className="flex-none md:flex">
          <div className="flex-1 grid grid-cols-2 gap-4">
            <ValueCard title="Total Reserve Assets" value={formatPrice(totalReserve)} />
            <ValueCard title="Total Minted IST" value={formatIST(totalMinted)} />
            <ValueCard title="Total Collateral Value Locked" value={formatPrice(totalLockedCollateral)} />
            <ValueCard title="IST minted by Vaults" value={formatIST(vaultMinted)} />
            <ValueCard title="Total PSM Assets" value={formatPrice(psmMinted)} />
            <ValueCard title="IST minted by PSM" value={formatIST(psmMinted)} />
            <ValueCard title="Reserve Shortfall" value={formatPrice(reserveShortfall)} className="col-span-2" />
          </div>
          <div className="w-[400px] ml-5">
            <PieChart width={400} height={400}>
              <Pie
                dataKey="value"
                data={[
                  {
                    name: 'Vaults',
                    value: roundPrice(vaultMinted),
                  },
                  {
                    name: 'PSM',
                    value: roundPrice(psmMinted),
                  },
                ]}
                outerRadius={100}
                fill="green"
                labelLine={false}
                label={renderCustomizedLabel}
              >
                <Cell fill={colors[0]} />
                <Cell fill={colors[1]} />
              </Pie>
              <Tooltip />
            </PieChart>
          </div>
        </div>
      </PageContent>
    </>
  );
}
