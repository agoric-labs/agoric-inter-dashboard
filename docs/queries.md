## Check integrity

```sql
with mm as (
   select min(block_height) mn, max(block_height) mx
     from `xxx.raw_cosmos_hub.blocks`
)
select avg(events.count / b.event_count) as status_events
     , avg(msgs.count / b.message_count) as status_msgs
     , avg(txs.count / b.transaction_count) as status_txs
     , count(block_height) / count(distinct block_height) as status_blocks
     , count(block_height) / (max(block_height) - min(block_height)) as status_missing_total
     , sum(events.count) event_count
     , sum(msgs.count) msg_count
     , sum(txs.count) tx_count
     , count(block_height) as total_count
     , (max(block_height) - min(block_height)) as expected_total
     , (max(block_time) - min(block_time)) / count(block_height) avg_build_block_time
     , (max(_sdc_batched_at) - min(_sdc_batched_at)) / count(block_height) avg_extract_block_time
     , max(block_height) as last_block_height
     , max(block_height - prev_block_height) as max_hole
from (
  select block_height
       , block_time
       , event_count
       , message_count
       , transaction_count
       , _sdc_batched_at
       , lag(block_height) over (order by block_height) as prev_block_height
    from `xxx.raw_cosmos_hub.blocks`
    where block_height between (select mn from mm) + 500 and (select mx from mm) - 500
) b
left join (
  select block_height, count(block_height) count
    from `xxx.raw_cosmos_hub.events`
   group by 1
) events using (block_height)
left join (
  select block_height, count(block_height) count
    from `xxx.raw_cosmos_hub.messages`
   group by 1
) msgs using (block_height)
left join (
  select block_height, count(block_height) count
    from `xxx.raw_cosmos_hub.transactions`
   group by 1
) txs using (block_height)
```

## Fill missing blocks

```
kubectl port-forward agoric-mainnet-tendermint-head-5f876fb589-cvt5c 3333:3333
```

```
with mm as (
   select min(block_height) mn, max(block_height) mx
     from `xxx.raw_cosmos_hub.blocks`
)
select array_to_string(array_agg('curl -d \'' || to_json_string(to_json(struct(earliest, latest))) || '\' http://localhost:3333'), '\n'), count(earliest)
  from (
    select next as earliest
         , least(next + 99, block_height - 1) as latest
     from (
      select block_height
           , lag(block_height) over (order by block_height) as prev_block_height
        from `xxx.raw_cosmos_hub.blocks`
       where block_height between (select mn from mm) + 500 and (select mx from mm) - 500
     )
     cross join unnest(generate_array(prev_block_height + 1, block_height - 1, 100)) next
     where block_height - prev_block_height > 1
     order by 1
  )
```
