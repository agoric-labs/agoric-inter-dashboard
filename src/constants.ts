export const GET_MODULE_ACCOUNTS_URL = 'https://main-a.api.agoric.net:443/cosmos/auth/v1beta1/module_accounts';

export const GET_ACCOUNT_BALANCE_URL = (address: string) => `https://main-a.api.agoric.net:443/cosmos/bank/v1beta1/balances/${address}`;

export const GET_INTERCHAIN_BALANCES_URL = 'https://rickhouse.agoric-core.workers.dev/inter-balances';

export const UIST_DENOMINATION = 'uist';

export const VBANK_RESERVE_ACCOUNT = 'vbank/reserve';

export const GRAPH_DAYS = 90;

export const SUBQUERY_URL = 'https://index-api.onfinality.io/sq/agoric-labs/agoric-mainnet-v2';

export const enum VAULT_STATES {
  ACTIVE = 'active',
  LIQUIDATED = 'liquidated',
  LIQUIDATING = 'liquidating',
  CLOSED = 'closed',
}
