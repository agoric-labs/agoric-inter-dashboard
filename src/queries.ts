import { getDateKey, range } from "./utils";

export const PSM_DASHBOARD_QUERY = `
query {
    _metadata {
        lastProcessedHeight
    }
    psmMetrics {
        nodes {
            blockHeight
            denom
            mintedPoolBalance
        }
    }
    psmGovernances {
        nodes {
            mintLimit
            denom
        }
    }
    boardAuxes {
        nodes {
            allegedName
            decimalPlaces
        }
    }
}`;

export const PSM_GRAPH_TOKENS_QUERY = `
query {
    psmMetrics {
        nodes {
            id
            denom
        }
    }
}`;

export const PSM_TOKEN_DAILY_MINT_QUERY = (tokens: Array<string>) => `
query {
    ${tokens?.map((token: string) => `
        ${token}: psmMetricsDailies(first: 90, filter: {denom: {equalTo: "${token}"}}, orderBy:DATE_KEY_DESC) {
        nodes {
                id
                denom
                dateKey
                blockTimeLast
                totalMintedProvidedLast
            }
        }`
)}
}`
const currentDate = new Date();
export const VAULTS_DASHBOARD_QUERY = `
query {
    vaultManagerMetrics {
        nodes {
            id
            numActiveVaults
            totalCollateral
            totalDebt
            liquidatingCollateralBrand
        }
    }
    oraclePrices {
        nodes {
            priceFeedName
            typeOutAmount
            typeInAmount
        }
    }
    boardAuxes {
        nodes {
            allegedName
            decimalPlaces
        }
    }
    vaultManagerGovernances {
        nodes {
            id
            debtLimit
            liquidationMarginNumerator
            liquidationMarginDenominator
        }
    }
    vaults(filter: {state: {equalTo: "active"}}) {
        nodes {
            id
            denom
            balance
            state
            debt
            blockTime
        }
        totalCount
    }
    oraclePriceDailies(filter: {
        dateKey: { 
            in: [${getDateKey(currentDate).key}, ${getDateKey(currentDate, 1).key}] 
        }
    }) {
        nodes {
            typeInName
            typeInAmountLast
            typeOutAmountLast
            dateKey
        }
    }
}`;

export const OPEN_VAULTS_NEXT_PAGES_QUERY = (pages: number) => `
query {
    ${range(pages).map(index => `v${index}:vaults(filter: {state: {equalTo: "active"}}, first: 100, offset:${(index + 1) * 100}) {
        nodes {
            id
            denom
            balance
            state
            debt
            blockTime
        }
    }`)}
}`

export const VAULTS_DAILY_METRICS_QUERY = (tokens: string[]) => `
query {
    ${tokens?.map((token) =>
    `${token}: vaultManagerMetricsDailies(first: 90, filter:{liquidatingCollateralBrand: {equalTo: "${token}"} }, orderBy:DATE_KEY_DESC)  {
        nodes {
            id
            dateKey
            liquidatingCollateralBrand
            blockTimeLast
            totalCollateralLast
            metricsCount
            totalDebtLast
        }
    }
    ${token}_oracle: oraclePriceDailies (first: 90, filter:{typeInName: {equalTo: "${token}"}}, orderBy:DATE_KEY_DESC ) {
        nodes {
            id
            dateKey
            blockTimeLast
            typeInName
            typeInAmountLast
            typeOutAmountLast
        }
    }`)}
}`

export const RESERVE_DASHBOARD_QUERY = `
query {
    reserveMetrics {
        nodes {
            shortfallBalance
            allocations {
                nodes {
                    id
                    denom
                    value
                }
            }
        }
    }
    oraclePrices {
        nodes {
            typeInName
            typeOutAmount
            typeInAmount
        }
    }
    boardAuxes {
        nodes {
            allegedName
            decimalPlaces
        }
    }
}`


export const RESERVE_GRAPH_TOKENS_QUERY = `
query {
    reserveMetrics {
        nodes {
            allocations {
                nodes {
                    id
                    denom
                }
            }
        }
    }
}`

export const RESERVE_DAILY_METRICS_QUERY = (tokens: string[], startDate: number) => `
query {
    ${tokens?.map((token) =>
    `${token}: reserveAllocationMetricsDailies (first: 90, filter: {denom: {equalTo: "${token}"}}, orderBy:DATE_KEY_DESC) {
        nodes {
            id
            blockTimeLast
            dateKey
            valueLast
            denom
        }
    }
    ${token}_oracle: oraclePriceDailies (first: 90, filter: {typeInName: {equalTo: "${token}"}}, orderBy:DATE_KEY_DESC) {
        nodes {
          dateKey
          blockTimeLast
          typeInName
          typeInAmountLast
          typeOutAmountLast
        }
    }
    ${token}_last: reserveAllocationMetricsDailies (first: 1, filter: {denom: {equalTo: "${token}"}, valueLast: {greaterThan: "0" }, dateKey: {lessThan: ${startDate}}}, orderBy:DATE_KEY_DESC) {
        nodes {
          id
          blockTimeLast
          dateKey
          valueLast
          denom
        }
    }`)}
}`

export const INTER_DASHBOARD_QUERY = `
query {
    psmMetrics {
        nodes {
            denom
            mintedPoolBalance
            anchorPoolBalance
            mintedPoolBalance
        }
    }
    psmGovernances {
        nodes {
            denom
            mintLimit
        }
    }
    vaultManagerMetrics {
        nodes {
            id
            liquidatingCollateralBrand
            liquidatingDebtBrand
            totalDebt
            totalCollateral
        }
    }
    vaultManagerGovernances {
        nodes {
            id
            debtLimit
        }
    }
    reserveMetrics {
        nodes {
            totalFeeBurned
            shortfallBalance
            allocations {
                nodes {
                    id
                    denom
                    value
                }
            }
        }
    }
    wallets {
        totalCount
    }
    oraclePrices {
        nodes {
            id
            typeInName
            typeOutName
            typeInAmount
            typeOutAmount
            priceFeedName
        }
    }
    boardAuxes {
        nodes {
            allegedName
            decimalPlaces
        }
    }
}`

export const LIQUIDATIONS_DASHBOARD = `
query {
    boardAuxes {
        nodes {
            allegedName
            decimalPlaces
        }
    }
    vaultManagerMetrics {
        nodes {
            id
            numLiquidationsCompleted
            liquidatingCollateralValue
            numActiveVaults
            numLiquidationsCompleted
            numLiquidationsAborted
            numLiquidatingVaults
        }
    }
    vaultLiquidations (filter: {state: {equalTo: "liquidated"}}) {
        nodes {
            id
            denom
            debt
            state
            balance
            blockTime
            oraclePrice
            vaultManagerGovernance
            currentState {
                id
                denom
                debt
                state
                balance
                blockTime
            }
            liquidatingState {
                id
                denom
                debt
                state
                balance
                blockTime
            }
        }
    }
}`;

export const VAULT_STATE_DAILIES_QUERY = `
query {
    vaultStatesDailies  ( first: 90, orderBy: ID_DESC, offset: 1) {
        nodes {
            id
            blockTimeLast
            active
            closed
            liquidated
            liquidating
            liquidatedClosed
        }
    }
}`
