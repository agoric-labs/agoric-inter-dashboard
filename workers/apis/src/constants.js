export const balancesKey = 'balances';
export const BASE_NODE_URL = 'https://main-a.api.agoric.net:443';
export const SUBQUERY_URL = 'https://index-api.onfinality.io/sq/agoric-labs/agoric-mainnet-v2';
