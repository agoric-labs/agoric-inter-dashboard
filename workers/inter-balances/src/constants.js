export const balancesKey = 'balances';
export const BASE_NODE_URL = 'https://main-a.api.agoric.net:443';
export const SUBQUERY_URL = 'https://api.subquery.network/sq/agoric-labs/mainnet__YWdvc';