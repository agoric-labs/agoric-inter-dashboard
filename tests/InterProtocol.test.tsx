import useSWR from 'swr';
import { render } from '@testing-library/react';
import { InterProtocol } from '@/pages/InterProtocol';
import { dashboardDataMock, interchainBalancesMock } from './__mocks__/InterProtocol';
import { INTER_DASHBOARD_QUERY } from '@/queries';
import { GET_INTERCHAIN_BALANCES_URL } from '@/constants';

jest.mock('swr', () => {
  const originalModule = jest.requireActual('@/utils');
  return {
    ...originalModule,
    default: jest.fn(),
  };
});

describe('InterProtocol Snapshot tests', () => {
  it('should match snapshot when data is loaded', async () => {
    (useSWR as any).mockImplementation((query: string) => {
      const mockData = {
        [INTER_DASHBOARD_QUERY]: dashboardDataMock,
        [GET_INTERCHAIN_BALANCES_URL]: interchainBalancesMock,
      }[query];
      return { data: mockData, error: null, isLoading: false };
    });

    const { container } = render(<InterProtocol />);
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when data is loading', async () => {
    (useSWR as any).mockImplementation(() => ({ data: undefined, error: null, isLoading: true }));

    const { container } = render(<InterProtocol />);
    expect(container).toMatchSnapshot();
  });
});
