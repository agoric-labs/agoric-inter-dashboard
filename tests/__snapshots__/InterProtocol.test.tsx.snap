// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InterProtocol Snapshot tests should match snapshot when data is loaded 1`] = `
<div>
  <h1
    class="flex items-center space-x-4 lg:space-x-6 border-b p-8 text-3xl font-bold tracking-tight"
  >
    Summary
  </h1>
  <div
    class="p-8"
  >
    <div
      class="grid grid-cols-2 xl:grid-cols-3 gap-4 mb-5"
    >
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
        >
          <h3
            class="tracking-tight text-sm font-medium"
          >
            IST in Circulation
          </h3>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div
            class="text-2xl font-bold"
            data-testid="inter-protocol-minted"
          >
            1.57M
          </div>
        </div>
      </div>
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
        >
          <h3
            class="tracking-tight text-sm font-medium"
          >
            Total Mint Limit
          </h3>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div
            class="text-2xl font-bold"
            data-testid="inter-protocol-mint-limit"
          >
            16.7M
          </div>
        </div>
      </div>
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
        >
          <h3
            class="tracking-tight text-sm font-medium"
          >
            Total Mint Limit Utilized
          </h3>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div
            class="text-2xl font-bold"
            data-testid="inter-protocol-mint-limit-utilized"
          >
            9.38%
          </div>
        </div>
      </div>
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
        >
          <h3
            class="tracking-tight text-sm font-medium"
          >
            Total Interchain IST
          </h3>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div
            class="text-2xl font-bold"
          >
            421.28K
          </div>
        </div>
      </div>
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
        >
          <h3
            class="tracking-tight text-sm font-medium"
          >
            % of Interchain IST
          </h3>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div
            class="text-2xl font-bold"
          >
            26.9%
          </div>
        </div>
      </div>
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
        >
          <h3
            class="tracking-tight text-sm font-medium"
          >
            Smart Wallets Provisioned
          </h3>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div
            class="text-2xl font-bold"
          >
            1651
          </div>
        </div>
      </div>
    </div>
    <h2
      class="text-2xl fontebold tracking-tight mb-5"
    >
      Balances
    </h2>
    <div
      class="flex-none md:flex"
    >
      <div
        class="flex-1 grid grid-cols-2 gap-4"
      >
        <div
          class="rounded-lg border bg-card text-card-foreground shadow-sm"
        >
          <div
            class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
          >
            <h3
              class="tracking-tight text-sm font-medium"
            >
              Total Reserve Assets
            </h3>
          </div>
          <div
            class="p-6 pt-0"
          >
            <div
              class="text-2xl font-bold"
              data-testid="inter-protocol-total-reserve-assets"
            >
              $71.75K
            </div>
          </div>
        </div>
        <div
          class="rounded-lg border bg-card text-card-foreground shadow-sm"
        >
          <div
            class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
          >
            <h3
              class="tracking-tight text-sm font-medium"
            >
              Total Minted IST
            </h3>
          </div>
          <div
            class="p-6 pt-0"
          >
            <div
              class="text-2xl font-bold"
              data-testid="inter-protocol-total-minted"
            >
              1.57M
            </div>
          </div>
        </div>
        <div
          class="rounded-lg border bg-card text-card-foreground shadow-sm"
        >
          <div
            class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
          >
            <h3
              class="tracking-tight text-sm font-medium"
            >
              Total Collateral Value Locked
            </h3>
          </div>
          <div
            class="p-6 pt-0"
          >
            <div
              class="text-2xl font-bold"
              data-testid="inter-protocol-total-collateral-value-locked"
            >
              $5.00M
            </div>
          </div>
        </div>
        <div
          class="rounded-lg border bg-card text-card-foreground shadow-sm"
        >
          <div
            class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
          >
            <h3
              class="tracking-tight text-sm font-medium"
            >
              IST minted by Vaults
            </h3>
          </div>
          <div
            class="p-6 pt-0"
          >
            <div
              class="text-2xl font-bold"
              data-testid="inter-protocol-ist-minted-vaults"
            >
              1.53M
            </div>
          </div>
        </div>
        <div
          class="rounded-lg border bg-card text-card-foreground shadow-sm"
        >
          <div
            class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
          >
            <h3
              class="tracking-tight text-sm font-medium"
            >
              Total PSM Assets
            </h3>
          </div>
          <div
            class="p-6 pt-0"
          >
            <div
              class="text-2xl font-bold"
              data-testid="inter-protocol-total-psm-assets"
            >
              $28,665,205,252.95M
            </div>
          </div>
        </div>
        <div
          class="rounded-lg border bg-card text-card-foreground shadow-sm"
        >
          <div
            class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
          >
            <h3
              class="tracking-tight text-sm font-medium"
            >
              IST minted by PSM
            </h3>
          </div>
          <div
            class="p-6 pt-0"
          >
            <div
              class="text-2xl font-bold"
              data-testid="inter-protocol-ist-minted-psm"
            >
              32.79K
            </div>
          </div>
        </div>
        <div
          class="rounded-lg border bg-card text-card-foreground shadow-sm col-span-2"
        >
          <div
            class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
          >
            <h3
              class="tracking-tight text-sm font-medium"
            >
              Reserve Shortfall
            </h3>
          </div>
          <div
            class="p-6 pt-0"
          >
            <div
              class="text-2xl font-bold"
              data-testid="inter-protocol-reserve-shortfall"
            >
              $7.84K
            </div>
          </div>
        </div>
      </div>
      <div
        class="w-[400px] ml-5"
      >
        <div
          class="recharts-wrapper"
          role="region"
          style="position: relative; cursor: default; width: 400px; height: 400px;"
        >
          <svg
            class="recharts-surface"
            cx="50%"
            cy="50%"
            height="400"
            viewBox="0 0 400 400"
            width="400"
          >
            <title />
            <desc />
            <defs>
              <clippath
                id="recharts1-clip"
              >
                <rect
                  height="390"
                  width="390"
                  x="5"
                  y="5"
                />
              </clippath>
            </defs>
            <g
              class="recharts-layer recharts-pie"
              tabindex="0"
            >
              <g
                class="recharts-layer"
              >
                <g
                  class="recharts-layer recharts-pie-sector"
                  tabindex="-1"
                />
                <g
                  class="recharts-layer recharts-pie-sector"
                  tabindex="-1"
                />
              </g>
            </g>
          </svg>
          <div
            class="recharts-tooltip-wrapper"
            role="dialog"
            style="transform: translate(undefinedpx, undefinedpx); pointer-events: none; visibility: hidden; position: absolute; top: 0px; left: 0px;"
            tabindex="-1"
          >
            <div
              class="recharts-default-tooltip"
              style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); border: 1px solid #ccc; white-space: nowrap;"
            >
              <p
                class="recharts-tooltip-label"
                style="margin: 0px;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`InterProtocol Snapshot tests should match snapshot when data is loading 1`] = `
<div>
  <h1
    class="flex items-center space-x-4 lg:space-x-6 border-b p-8 text-3xl font-bold tracking-tight"
  >
    Summary
  </h1>
  <div
    class="p-8"
  >
    <div
      class="grid grid-cols-2 xl:grid-cols-3 gap-4 mb-5"
    >
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
        >
          <h3
            class="tracking-tight text-sm font-medium"
          >
            IST in Circulation
          </h3>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div
            class="text-2xl font-bold"
          >
            <div
              class="animate-pulse bg-muted w-[100px] h-[32px] rounded-full"
            />
          </div>
        </div>
      </div>
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
        >
          <h3
            class="tracking-tight text-sm font-medium"
          >
            Total Mint Limit
          </h3>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div
            class="text-2xl font-bold"
          >
            <div
              class="animate-pulse bg-muted w-[100px] h-[32px] rounded-full"
            />
          </div>
        </div>
      </div>
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
        >
          <h3
            class="tracking-tight text-sm font-medium"
          >
            Total Mint Limit Utilized
          </h3>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div
            class="text-2xl font-bold"
          >
            <div
              class="animate-pulse bg-muted w-[100px] h-[32px] rounded-full"
            />
          </div>
        </div>
      </div>
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
        >
          <h3
            class="tracking-tight text-sm font-medium"
          >
            Total Interchain IST
          </h3>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div
            class="text-2xl font-bold"
          >
            <div
              class="animate-pulse bg-muted w-[100px] h-[32px] rounded-full"
            />
          </div>
        </div>
      </div>
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
        >
          <h3
            class="tracking-tight text-sm font-medium"
          >
            % of Interchain IST
          </h3>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div
            class="text-2xl font-bold"
          >
            <div
              class="animate-pulse bg-muted w-[100px] h-[32px] rounded-full"
            />
          </div>
        </div>
      </div>
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
        >
          <h3
            class="tracking-tight text-sm font-medium"
          >
            Smart Wallets Provisioned
          </h3>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div
            class="text-2xl font-bold"
          >
            <div
              class="animate-pulse bg-muted w-[100px] h-[32px] rounded-full"
            />
          </div>
        </div>
      </div>
    </div>
    <h2
      class="text-2xl fontebold tracking-tight mb-5"
    >
      Balances
    </h2>
    <div
      class="animate-pulse bg-muted w-full h-[20px] rounded-full mb-2"
    />
    <div
      class="animate-pulse bg-muted w-full h-[20px] rounded-full mb-2"
    />
    <div
      class="animate-pulse bg-muted w-full h/[20px] rounded-full mb-2"
    />
    <div
      class="animate-pulse bg-muted w-full h-[20px] rounded-full mb-2"
    />
  </div>
</div>
`;
